sdk/storage/inc/common/xml_wrapper.hpp

```diff
@@ -0,0 +1,285 @@
+  /// Adds a namespace declaration to the node.
+  /// </summary>
+  /// <param name="uri">The namespace to associate with the prefix.</param>
+  /// <param name="prefix">The namespace prefix</param>
+  void set_namespace_declaration(const std::string& uri, const std::string& prefix);
+
+  /// <summary>
+  /// Set the namespace prefix
+  /// </summary>
+  /// <param name="prefix">name space prefix to be set</param>
+  void set_namespace(const std::string& prefix);
+
+  /// <summary>
+  /// Sets the value of the attribute with this name (and prefix).
+  /// </summary>
+  /// <param name="name">The name of the attribute</param>
+  /// <param name="value">The value of the attribute</param>
+  /// <param name="prefix">The prefix of the attribute, this is optional.</param>
+  void set_attribute(const std::string& name, const std::string& value, const std::string& prefix);
+
+  /// <summary>
+  /// Sets the text of the first text node. If there isn't a text node, add one and set it.
+  /// </summary>
+  /// <param name="text">The text to be set to the child node.</param>
+  void set_child_text(const std::string& text);
+
+  /// <summary>
+  /// Frees the wrappers set in nod->_private
+  /// </summary>
+  /// <param name="node">The node to be freed.</param>
+  /// <returns>A <see cref="pplx::task" /> object that represents the current operation.</returns>
+  static void free_wrappers(xmlNode* node);
+
+private:
+  xmlNode* m_ele;
+};
+
+/// <summary>
+/// A class to wrap xmlDoc of c library libxml2. This class provides abilities to create xml format
+/// texts from nodes.
+/// </summary>
+class xml_document_wrapper
+{
+public:
+  xml_document_wrapper();
+
+  ~xml_document_wrapper();
+
+  /// <summary>
+  /// Converts object to a string object.
+  /// </summary>
+  /// <returns>A std::string that contains the result</returns>
+  std::string write_to_string();
+
+  /// <summary>
+  /// Creates the root node of the document.
+  /// </summary>
+  /// <param name="name">The name of the root node.</param>
+  /// <param name="namespace_name">The namespace of the root node.</param>
+  /// <param name="prefix">The namespace prefix of the root node.</param>
+  /// <returns>A wrapper that contains the root node.</returns>
+  xml_element_wrapper* create_root_node(
+      const std::string& name,
+      const std::string& namespace_name,
+      const std::string& prefix);
+
+  /// <summary>
+  /// Gets the root node of the document.
+  /// </summary>
+  /// <returns>The root node of the document.</returns>
+  xml_element_wrapper* get_root_node() const;
+
+private:
+  xmlDocPtr m_doc;
+};
+
+/// <summary>
+/// XML writer based on libxml2
+/// </summary>
+class xml_writer
+{
+public:
+  virtual ~xml_writer() {}
+
+protected:
+  xml_writer() : m_stream(nullptr)
+  {
+  }
+
+  /// <summary>
+  /// Initialize the writer
+  /// </summary>
+  void initialize(std::ostream& stream);
+
+  /// <summary>
+  /// Finalize the writer
+  /// </summary>
+  void finalize();
```

Comment:
What does this even mean? Finalizers are not a C++ concept, we have destructors.